ash/src/vk.rs

```diff
@@ -49,11 +49,83 @@ pub trait Handle: Sized {
     }
 }
 
+/// Iterates through the pointer chain. Includes the item that is passed into the function. Stops at
+/// the last [`BaseOutStructure`] that has a null [`BaseOutStructure::p_next`] field.
+pub(crate) unsafe fn ptr_chain_iter<'a, T: TaggedStructure<'a>>(
+    ptr: &mut T,
+) -> impl Iterator<Item = *mut BaseOutStructure<'_>> {
+    let ptr = <*mut T>::cast::<BaseOutStructure<'_>>(ptr);
+    (0..).scan(ptr, |p_ptr, _| {
+        if p_ptr.is_null() {
+            return None;
+        }
+        let n_ptr = (**p_ptr).p_next;
+        let old = *p_ptr;
+        *p_ptr = n_ptr;
+        Some(old)
+    })
+}
+
 /// Structures implementing this trait are layout-compatible with [`BaseInStructure`] and
 /// [`BaseOutStructure`]. Such structures have an `s_type` field indicating its type, which must
 /// always match the value of [`TaggedStructure::STRUCTURE_TYPE`].
-pub unsafe trait TaggedStructure {
+pub unsafe trait TaggedStructure<'a>: Sized {
```

Comment:
Is `: Sized` necessary here? I think it's more conventional to only state this explicitly if removing it would break downstream code.

Reply:
Not sure if it was missed, but I posed this as a question (to you) in my message above:

> The presence of `const STRUCTURE_TYPE` on `TaggedStructure` removes object safety, but Rust only complains about the `+ ?Sized` bound being useless when adding an explicit `TaggedStructure: Sized` requirement...

In my eyes these are similar. Is the linter simply not strong enough to detect this case, including:

> It says nothing about the `where Self: Sized` bound being redundant either way.

I can remove it either way.